core/src/main/java/feign/AsyncResponseHandler.java

```diff
@@ -0,0 +1,102 @@
+
+package feign;
+
+import static feign.FeignException.*;
+import static feign.Util.*;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.concurrent.CompletableFuture;
+
+import feign.Logger.Level;
+import feign.codec.DecodeException;
+import feign.codec.Decoder;
+import feign.codec.ErrorDecoder;
+
+/**
+ * The response handler that is used to provide asynchronous support on top of standard
+ * response handling
+ */
+class AsyncResponseHandler {
+
+	private static final long MAX_RESPONSE_BUFFER_SIZE = 8192L;
+
+	private final Level logLevel;
+	private final Logger logger;
+
+	private final Decoder decoder;
+	private final ErrorDecoder errorDecoder;
+	private final boolean decode404;
+	private final boolean closeAfterDecode;
+
+	AsyncResponseHandler(Level logLevel, Logger logger, Decoder decoder, ErrorDecoder errorDecoder,
+			boolean decode404, boolean closeAfterDecode) {
+		super();
+		this.logLevel = logLevel;
+		this.logger = logger;
+		this.decoder = decoder;
+		this.errorDecoder = errorDecoder;
+		this.decode404 = decode404;
+		this.closeAfterDecode = closeAfterDecode;
+	}
+	
+	void handleResponse(CompletableFuture<Object> resultFuture, AsyncInvocation<?> invocationContext,
```

Comment:
There is a large amount of duplication here from the regular method handler.  We should consider refactoring this logic into a component that can be used by both execution modes.

Reply:
Yes, there is. I first wanted to bring this in cleanly for review without doing major changes to unify everything.

Reply:
Refactored.